src/Sylius/Component/Inventory/Operator/BackordersHandler.php

```diff
@@ -78,16 +86,19 @@ public function fillBackorders(StockableInterface $stockable)
             return;
         }
 
-        $units = $this->repository->findBy([
-            'stockable' => $stockable,
-            'inventoryState' => InventoryUnitInterface::STATE_BACKORDERED,
-        ], ['createdAt' => 'ASC']);
+        $units = $this->repository->findByStockableAndInventoryState(
+            $stockable,
+            InventoryUnitInterface::STATE_BACKORDERED,
+            $onHand
+        );
 
         foreach ($units as $unit) {
-            $unit->setInventoryState(InventoryUnitInterface::STATE_SOLD);
+            $sm = $this->factory->get($unit, InventoryUnitTransitions::GRAPH);
 
-            if (--$onHand === 0) {
-                break;
+            if ($sm->apply(InventoryUnitTransitions::SYLIUS_HOLD)) {
```

Comment:
@lchrusciel Should I use the `soft` transition here? Or do you want an exception if the transition can't be applied?